### ISSUE TITLE
Failing to drop `HeaderMap::Drain` causes double-free

### ISSUE THREAD

**Issue by Qwaz** [hyperium/http#354]
https://github.com/hyperium/http/blob/9c05e391e00474abaa8c14a86bcb0fc5eff1120e/src/header/map.rs#L2115-L2122

https://github.com/hyperium/http/blob/9c05e391e00474abaa8c14a86bcb0fc5eff1120e/src/header/map.rs#L2140-L2148

[Failing to drop a value is considered safe](https://doc.rust-lang.org/nomicon/leaking.html) in Rust, and unsafe code should not rely on this behavior.

> It is reasonable for safe code to assume that destructor leaks do not happen, as any program that leaks destructors is probably wrong. However unsafe code cannot rely on destructors to be run in order to be safe.

`HeaderMap::Drain` uses `ptr::read` to move out entries from the map when it iterates, and calls `map.entries.set_len(0)` to clear the map at once when it is dropped. If `Drain`'s drop is not called, double-free happens when `HeaderMap` is dropped. Also, if `Drain` is dropped without iterating to the end, it leaks memory.

[Demonstration](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c5d2488bd0ef3ea6826cc5b17d264845)

**Comment by Shnatsel** [comment]
Double free is an [exploitable security vulnerability](https://heap-exploitation.dhavalkapil.com/attacks/double_free.html). After releasing a fixed version please file a [security advisory](https://github.com/RustSec/advisory-db) and, if possible, yank the affected versions from crates.io.

**Comment by hawkw** [comment]
I started attempting to fix this using a similar approach as [`Vec::drain`](https://doc.rust-lang.org/nomicon/leaking.html#drain), setting the length of `self.entries` to 0 when creating the `Drain` iterator, and then setting the length back if the `Drain` iterator is dropped before draining all elements. This way, `mem::forget`ting the iterator would only amplify a leak, rather than calling a double-free. 

However, this isn't trivially possible here: the `ValueDrain` iterator calls `HeaderMap::remove_extra_values`, which relies on indexing the `entries` vector. If we've already set it's length to 0, the bounds checks there will fail. So, either that method needs to be refactored to not rely on indexing, or we need a different solution here (possibly involving `ManuallyDrop`?). I'll keep working on it.
